pkg/utils/baremetal/bootstrap.go

```diff
@@ -0,0 +1,94 @@
+package baremetal
+
+import (
+	"context"
+	"fmt"
+
+	baremetalhost "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	clientwatch "k8s.io/client-go/tools/watch"
+)
+
+// WaitForBaremetalBootstrapControlPlane will watch baremetalhost resources on the bootstrap
+// and wait for the control plane to finish provisioning.
+func WaitForBaremetalBootstrapControlPlane(ctx context.Context, config *rest.Config) error {
+	client, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("creating a baremetal client: %w", err)
+	}
+
+	r := client.Resource(baremetalhost.GroupVersion.WithResource("baremetalhosts")).Namespace("openshift-machine-api")
+	blw := BmhCacheListerWatcher{
+		Resource: r,
+	}
+
+	logrus.Infof("  Waiting for baremetal control plane to provision...")
+
+	masters := map[string]baremetalhost.BareMetalHost{}
+
+	_, err = clientwatch.UntilWithSync(
+		ctx,
+		blw,
+		&unstructured.Unstructured{},
+		nil,
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Added, watch.Modified:
+			default:
+				return false, nil
+			}
+
+			bmh := &baremetalhost.BareMetalHost{}
+
+			unstr, err := runtime.DefaultUnstructuredConverter.ToUnstructured(event.Object)
+			if err != nil {
+				return false, err
+			}
+
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstr, bmh); err != nil {
+				logrus.Error("failed to convert to bmh", err)
+				return false, err
+			}
+
+			role, found := bmh.Labels["installer.openshift.io/role"]
+
+			if found && role == "control-plane" {
+				prev, found := masters[bmh.Name]
+
+				if !found || bmh.Status.Provisioning.State != prev.Status.Provisioning.State {
+					if bmh.Status.Provisioning.State == baremetalhost.StateNone {
+						// StateNone is an empty string
+						logrus.Infof("  baremetalhost: %s: uninitialized", bmh.Name)
+					} else {
+						logrus.Infof("  baremetalhost: %s: %s", bmh.Name, bmh.Status.Provisioning.State)
+					}
+
+					if bmh.Status.OperationalStatus == baremetalhost.OperationalStatusError {
+						logrus.Warnf("  baremetalhost: %s: error: %s %s", bmh.Name, bmh.Status.ErrorType, bmh.Status.ErrorMessage)
```

Comment:
ErrorTypes are [all of the form](https://github.com/openshift/baremetal-operator/blob/master/apis/metal3.io/v1alpha1/baremetalhost_types.go#L159-L182) `<x> error`, so I think we could shorten `error: %s` to just `%s:`.

Reply:
Done